string_grouper/test/test_string_grouper.py

```diff
@@ -124,6 +124,32 @@ def test_match_strings(self, mock_StringGouper):
         mock_StringGrouper_instance.get_matches.assert_called_once()
         assert df == 'expected_result'
 
+    @patch('string_grouper.string_grouper.StringGrouper._symmetrize_matches_list')
```

Comment:
Part of the goal of a unittest is also to document the code and the reason of it. For me it is still not clear why we need the _symmetrize_matches_list (e.g. what the problem is it is trying to solve). I think a minimal test (that doesn't read in the entire CSV file) would be better. Also, maybe it is late but I don't see the difference between the results of the `test_match_list_symmetry_without_symmetrize_function` and the `test_match_list_symmetry_with_symmetrize_function`.

Another minor detail: I think I've been using the self.assert_equal function of the unittest framework in the other test. For consistency it would be good too use this as well.

Reply:
Hi @Bergvca ,

> Many thanks again! I haven't looked much yet at the removal of the reset index code yet, but on first glance it looks good! I've added a comment on the new unittest for symmetrize matches.

Glad you like it!  At the moment I'm wondering if for completeness I should extend the index functionality to the other 2 high-level functions: `group_similar_strings` and `match_most_similar` so that the string indexes can appear as a column in the result even if they are not explicitly passed as arguments to the functions.  Or are they OK as they currently stand?  Let me know what you think.

> Part of the goal of a unittest is also to document the code and the reason of it. 

It escaped me.  I've added the comments now.

> I don't see the difference between the results of the 
> test_match_list_symmetry_without_symmetrize_function and the 
> test_match_list_symmetry_with_symmetrize_function.

Sorry, I shouldn't have included the `max_n_matches` keyword argument-setting in the call to `match-strings`.  Actually it was meant for you, since [you suggested it was the culprit](https://github.com/Bergvca/string_grouper/pull/34#discussion_r593951683). And you were right, as I later confirmed, and I must have forgotten to remove it.  Indeed, when the option is not set `_matches_list` becomes non-symmetric and the unit-test fails.

>  I think a minimal test (that doesn't read in the entire CSV file) would be better. 

I agree.  I only meant to show you how I first discovered the bug when I didn't know why it occurred. I will amend the unit-test to use a minimal sample dataset.  

> For me it is still not clear why we need the _symmetrize_matches_list (e.g. what the problem is it is trying to solve).

I suppose it is chiefly a matter of consistency.  

Strictly speaking, the symmetry of `StringGrouper._matches_list` need not be a requirement.  
But my feeling is that in the case of `duplicates=None` it is often expected.  
In fact, you might recall [it came as a surprise to us both](https://github.com/Bergvca/string_grouper/pull/34#discussion_r593951683) when it turned out not to be the case.  
And when I first wrote the code to compute the centroid, I assumed it was symmetric, and consequently got the wrong result for the centroid.  
Moreover, it seems somewhat inconsistent that some matches would be accompanied by their converses while others would not, with the user not being able to control which ones will be symmetrized and which ones will not.  Anyway, I'll leave the decision up to you.

To remove the arbitrariness of the situation we could choose either to completely symmetrize the list, as I have now done, or completely “unsymmetrize” it (in which case each match appears once and only once without its converse).

Reply:
> Glad you like it! At the moment I'm wondering if for completeness I should extend the index functionality to the other 2 high-level functions: group_similar_strings and match_most_similar so that the string indexes can appear as a column in the result even if they are not explicitly passed as arguments to the functions. Or are they OK as they currently stand? Let me know what you think.

Yes it does make sense, although we are kind of "over-complete" because we also have the optional ID column. But for completeness I think it is nicer. The user can either change their ID column to the index and not use the optional ID parameter, or use the ID parameter, but still be aware of the original indices. 

> I suppose it is chiefly a matter of consistency.

Strictly speaking, the symmetry of StringGrouper._matches_list need not be a requirement.
But my feeling is that in the case of duplicates=None it is often implicitly expected.
In fact, you might recall it came as a surprise to us both when it turned out not to be the case.
And when I first wrote the code to compute the centroid, I assumed it was symmetric, and consequently got the wrong result for the centroid.
Moreover, it seems somewhat inconsistent that some matches would be accompanied by their converses while others would not, with the user not being able to control which ones will be symmetrized and which ones will not. Anyway, I'll leave the decision up to you.

Sorry I was too brief - I understood the purpose of the function, but not by reading the unittest :). 

I think it needs to be completely symmetrical. For example, if you want to see all matches of a specific string, you want to filter on the first column, and see all matches in the second column. If it is completely "asymmetrical"  you would have to filter both columns, and look at both. 

Anyway - I think this PR looks good to merge! Do you still want to add anything or can merge it?

Reply:
Please go ahead and merge.  I can always create a new PR for the outstanding code.